keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1503 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With this enhancement, Kubernetes Developers will declare validation rules using
+IDL tags in the `types.go` files that to define the Kubernetes native API types.
+For example:
+
+```go
+// staging/src/k8s.io/api/core/v1/types.go
+
+// +valdiationRule="!self.hostNetwork || self.containers.all(c, c.containerPort.all(cp, cp.hostPort == cp.containerPort))"
+type PodSpec struct {
+  Containers []Container `json:...`
+}
+
+type Container struct {
+  // ...
+
+  Ports []ContainerPort `json:...`
+
+  // ...
+}
+
+type ContainerPort struct {
+  // ...
+
+  // ...
+  //
+  // +minimum=1
+  // +maximum=65535
+  HostPort int32 `json:...`
+
+  // ...
+  //
+  // +minimum=1
+  // +maximum=65535
+  ContainerPort int32 `json:...`
+
+  // ...
+}
+```
+
+In this example, both `+valdiation`, `+minimum` and `+maximum` are IDL tags.
+
+The declarative validation rules will be used by the kube-apiserver to
+validate API requests.
+
+The declarative validation rules will also be included in the published OpenAPI:
+
+```json
+// Pod Spec
+"openAPIV3Schema": {
+  "type": "object",
+  "x-kubernetes-validations": [
+    {
+      "rule": "!self.hostNetwork || self.containers.all(c, c.containerPort.all(cp, cp.hostPort == cp.containerPort))"
```

Comment:
I think that's fine, but why is this a map with a `rule` field rather than just a list of rules?

Reply:
Somewhat similar to the problem with constants we've had for defaults, will I be able to refer to constants in the CEL expression?

Reply:
> I think that's fine, but why is this a map with a rule field rather than just a list of rules?
> regarding being a map

It's actually a list of maps, so you can specify multiple rules already. It's a map because a given rule contains the expression for the rule, and optionally a message for the error, field path, and error reason code

> Will it be possible for a field to have multiple +validation lines logically ANDed

specifying multiple rules will AND them together. If you are asking if one rule can be ANDED across multiple lines I think we can address that when we allow multi-line CEL validations